### PR TITLE
Improve code coverage

### PR DESCRIPTION
**Pull request by peter279k**
# Changed log
- improve code coverage.(refer this code coverage [report](https://coveralls.io/github/peter279k/process))
- add the more tests.

### CODE REVIEW

File: test/ProcessTest.php [hunk 1]
```diff
@@ -59,6 +65,186 @@ public function testCommandCanRun() {
         });
     }
 
+    public function testCommandArray() {
+        Loop::run(function () {
+            $process = new Process([self::CMD_PROCESS]);
```
Comment (kelunik) [inline test/ProcessTest.php]: Please use `["echo", "foo"]` there.

File: test/ProcessTest.php [hunk 2]
```diff
@@ -59,6 +65,186 @@ public function testCommandCanRun() {
...
+        $process = new Process([self::CMD_PROCESS], null, [
+            ['error_value']
+        ]);
+    }
+
+    /**
+     * @expectedException \Amp\Process\StatusError
+     * @expectedExceptionMessage The process has not been started
+     */
+    public function testGetStdinIsNull() {
```
Comment (kelunik) [inline test/ProcessTest.php]: This actually tests it throws, so the test name should be renamed accordingly.

**Comment by peter279k** [comment]
Hi @kelunik, thank you for your review.
I have pushed the latest commit.
Please review this.

Thanks.

File: test/ProcessTest.php [hunk 3]
```diff
@@ -67,7 +67,7 @@ public function testCommandCanRun() {
 
     public function testCommandArray() {
         Loop::run(function () {
-            $process = new Process([self::CMD_PROCESS]);
+            $process = new Process(['echo', 'foo']);
```
Comment (kelunik) [inline test/ProcessTest.php]: This should actually be applied everywhere. An array is used to build the command from its arguments, but `self::CMD_PROCESS` is already a full command string.

**Comment by peter279k** [comment]
Ok, I've modified the CMD_PROCESS variable value globally.

Thank you for this review.

File: test/ProcessTest.php [hunk 4]
```diff
@@ -59,6 +65,186 @@ public function testCommandCanRun() {
...
+        Loop::run(function () {
+            $process = new Process(self::CMD_PROCESS, __DIR__);
+            $this->assertSame(__DIR__, $process->getWorkingDirectory());
+        });
+    }
+
+    public function testGetEnv() {
+        Loop::run(function () {
+            $process = new Process(self::CMD_PROCESS, null, []);
+            $this->assertCount(0, $process->getEnv());
```
Comment (kelunik) [inline test/ProcessTest.php]: That can be `assertSame([], $process->getEnv())`.

File: test/ProcessTest.php [hunk 5]
```diff
@@ -2,12 +2,18 @@
...
 use Amp\Loop;
 use Amp\Process\Process;
 use PHPUnit\Framework\TestCase;
 
 class ProcessTest extends TestCase {
-    const CMD_PROCESS = 'echo foo';
+    const CMD_PROCESS = ['echo', 'foo'];
+
+    protected function setUp() {
+        file_exists(__DIR__.'/../stream') ? null:touch(__DIR__.'/../stream');
```
Comment (kelunik) [inline test/ProcessTest.php]: Please add spaces like so: `null : touch(...)`

File: test/ProcessTest.php [hunk 6]
```diff
@@ -59,6 +65,186 @@ public function testCommandCanRun() {
...
+
+    public function testGetStdinIsCustomized() {
+        Loop::run(function () {
+            $process = new Process(self::CMD_PROCESS, null, [], [
+                new ResourceInputStream(fopen(__DIR__.'/../stream', 'r')),
+            ]);
+            $process->start();
+            $promise = $process->join();
+            $completed = false;
+            $promise->onResolve(function () use (&$completed) { $completed = true; });
```
Comment (kelunik) [inline test/ProcessTest.php]: `$completed` is unused here.

File: test/ProcessTest.php [hunk 7]
```diff
@@ -59,6 +65,186 @@ public function testCommandCanRun() {
...
+
+    public function testGetStdoutIsCustomized() {
+        Loop::run(function () {
+            $process = new Process(self::CMD_PROCESS, null, [], [
+                new ResourceOutputStream(fopen(__DIR__.'/../stream', 'w')),
+            ]);
+            $process->start();
+            $promise = $process->join();
+            $completed = false;
+            $promise->onResolve(function () use (&$completed) { $completed = true; });
```
Comment (kelunik) [inline test/ProcessTest.php]: `$completed` is unused here.

File: test/ProcessTest.php [hunk 8]
```diff
@@ -59,6 +65,186 @@ public function testCommandCanRun() {
...
+
+    public function testGetStderrIsCustomized() {
+        Loop::run(function () {
+            $process = new Process(self::CMD_PROCESS, null, [], [
+                new ResourceOutputStream(fopen(__DIR__.'/../stream', 'w')),
+            ]);
+            $process->start();
+            $promise = $process->join();
+            $completed = false;
+            $promise->onResolve(function () use (&$completed) { $completed = true; });
```
Comment (kelunik) [inline test/ProcessTest.php]: `$completed` is unused here.

File: test/ProcessTest.php [hunk 9]
```diff
@@ -59,6 +65,186 @@ public function testCommandCanRun() {
...
+        $process->start();
+        $promise = $process->join();
+        $completed = false;
+        $promise->onResolve(function () use (&$completed) { $completed = true; });
+        $this->assertSame('env_value', $process->getEnv()[0]);
+    }
+
+    public function testProcessDeferredIsFailed() {
+        $process = new Process(new \Exception());
+        $this->assertNull($process->start());
```
Comment (kelunik) [inline test/ProcessTest.php]: This doesn't seem to test what the test name says.

File: test/ProcessTest.php [hunk 10]
```diff
@@ -59,6 +65,186 @@ public function testCommandCanRun() {
...
+    /**
+     * @expectedException \Amp\Process\StatusError
+     * @expectedExceptionMessage The process has not been started
+     */
+    public function testProcessCanReset() {
+        $process = new Process(self::CMD_PROCESS);
+        $process->start();
+        $promise = $process->join();
+        $completed = false;
+        $promise->onResolve(function () use (&$completed) { $completed = true; });
```
Comment (kelunik) [inline test/ProcessTest.php]: `$completed` is unused here.

File: test/ProcessTest.php [hunk 11]
```diff
@@ -59,6 +65,186 @@ public function testCommandCanRun() {
...
+        $processReset = clone $process;
+        $processReset->getPid();
```
Comment (kelunik) [inline test/ProcessTest.php]: Better use `$this->expectException` right above this line.

File: test/ProcessTest.php (hunk 9 above)
Comment (peter279k) [reply test/ProcessTest.php]: Ok....This is my testing method and I forget to remove it.

**Comment by kelunik** [comment]
Sorry for not merging this earlier. Could you rebase your changes onto the current master branch and fix the tests for the minor BC breaks mentioned in the [release notes](https://github.com/amphp/process/releases/tag/v0.3.0)?

**Comment by peter279k** [comment]
Hi @kelunik , thank you for your reply.
I will do this and I'm sorry for that I'm delay for this notification.

Thanks.

**Comment by peter279k** [comment]
According to this [issue](https://github.com/composer/composer/issues/6859), just wait for the Composer fix this bug. Then the Travis build will be worked successfully.

File: test/ProcessTest.php (hunk 10 above)
Comment (kelunik) [reply test/ProcessTest.php]: ping, there's no assertion for `$completed`, please either remove it or add an assertion.

File: lib/Process.php [hunk 12]
```diff
@@ -238,7 +238,7 @@ public function getStdout(): ProcessInputStream {
      */
     public function getStderr(): ProcessInputStream {
         if (!$this->handle) {
-            throw new StatusError("The process has not been started");
+            throw new StatusError("Process has not been started.");
```
Comment (kelunik) [inline lib/Process.php]: Why did you change these?

File: test/ProcessTest.php [hunk 13]
```diff
@@ -52,13 +56,168 @@ public function testCommandCanRun() {
...
+
+    /**
+     * @expectedException \Error
+     * @expectedExceptionMessage Cloning is not allowed!
+     */
+    public function testProcessResetSignalIsNotRunning() {
+        $process = new Process(self::CMD_PROCESS);
+        $process->start();
+        $promise = $process->join();
+        $promise->onResolve(function () use (&$completed) { $completed = true; });
```
Comment (kelunik) [inline test/ProcessTest.php]: If you remove the `$completed` assertions, you can remove this line as well.

**Comment by peter279k** [comment]
When I track the Travis build during testing PHP 7.0, the options-resolver version which is the PHP-CS-Fixer dependencies is not correct.
And I remove the ```--ignore-platform-reqs``` option, the Travis build is successful.
I think this [PR](https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/2722) is not problem at this time so I remove this option.

File: lib/Process.php (hunk 12 above)
Comment (peter279k) [reply lib/Process.php]: I think the exception message need to be unified.
Some place using ```Pocess``` and others using ```The process```.

**Comment by trowski** [comment]
@peter279k Removing `--ignore-platform-reqs` is not really correct, as that causes the nightly build to fail. We're trying to decide what to do with the travis build so we don't have to keep changing it due to PHP-CS-Fixer. I'll just remove that change when I squash this.

**Comment by peter279k** [comment]
Hi @trowski, thank you for your explanation.
When adding the ```ignore-platform-reqs``` option, the ```option-resolver``` dependency version will not install the correct version.

The ```option-resolver``` version is ```3.8.x``` in the PHP 7.0, not ```4.0.x```.
You can look at this [log](https://travis-ci.org/peter279k/process/jobs/310076432) and this removed option [log](https://travis-ci.org/peter279k/process/jobs/310111681).
It seems that this option will force to install the specific version, even the environment doesn't satisfy the requirements.

Thanks.

**Comment by trowski** [comment]
@peter279k Right, but we still need `--ignore-platform-reqs` for the nightly build. I'm considering making the composer install line conditional on the build. So something like:

```bash
  # --ignore-platform-reqs, because https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/2722
  - if [ "$TRAVIS_PHP_VERSION" == "nightly" ]; then
      composer update -n --prefer-dist --ignore-platform-reqs;
    else
      composer update -n --prefer-dist;
    fi

```

**Comment by peter279k** [comment]
@trowski, I think it's the good idea for making the different composer command conditional.

**Comment by trowski** [comment]
Manually merged via 2a429681bde02fb0664ea7166e0d6e4bc31e3d39. Thanks!
